project/npda/forms/async_model_form.py

```diff
@@ -0,0 +1,104 @@
+from django.forms import ModelForm, ValidationError
+from django.forms.utils import ErrorDict
+
+from asgiref.sync import async_to_sync
+
+import httpx
+
+"""
+Opt in to support form cleaning by making async HTTP requests with httpx. 
+
+To define an async per field cleaner:
+    async def aclean_field_name(self, async_client):
+
+You can mix async and sync per field cleaners and all will be run.
+
+To define an async form .clean:
+    async def aclean(self, async_client):
+
+We have to pass the async_client as a parameter otherwise it will stop working on
+subsequent requests as they can't be re-used across event loops and we are still running
+under WSGI so have a new async event loop for each request.
+
+You can't have both an async and a sync clean method for a given field or the whole form,
+the async one will take precendence.
+
+You can call afull_clean() to run them all asynchronously or use the existing Django
+API calling .errors, .is_valid to run them synchronously. The class keeps track of
+what has been run so regardless of how you call it, they will only run once.
+
+This class mirrors the internals of ModelForm so could break across versions.
+That said, many people override the methods we mirror so they're unlikely to change.
+However you should unit test anything using this to ensure it works across updates.
+"""
+class AsyncModelForm(ModelForm):
```

Comment:
I'm not sure if this abstraction is a good idea, originally I didn't have to pass async_client to each method because I stored it in a contextvar. This was much nicer but I hit issues where the async_client couldn't be used again in a different event loop so the first CSV upload or form submit worked but then subsequent ones failed

Reply:
yeah, canned this as it broke checking NHS number was not null. I don't know why and that demonstrates that this approach was digging too deep overriding Django internals